packages/nx/src/command-line/release/version.ts

```diff
@@ -0,0 +1,428 @@
+import { relative } from 'node:path';
+import {
+  // @ts-ignore missing from @types/semver
+  RELEASE_TYPES,
+  valid,
+} from 'semver';
+import { Generator } from '../../config/misc-interfaces';
+import { readNxJson } from '../../config/nx-json';
+import {
+  ProjectGraph,
+  ProjectGraphProjectNode,
+} from '../../config/project-graph';
+import {
+  NxJsonConfiguration,
+  joinPathFragments,
+  logger,
+  output,
+  workspaceRoot,
+} from '../../devkit-exports';
+import { FsTree, Tree, flushChanges } from '../../generators/tree';
+import {
+  createProjectGraphAsync,
+  readProjectsConfigurationFromProjectGraph,
+} from '../../project-graph/project-graph';
+import { NxArgs } from '../../utils/command-line-utils';
+import { findMatchingProjects } from '../../utils/find-matching-projects';
+import { combineOptionsForGenerator } from '../../utils/params';
+import { parseGeneratorString } from '../generate/generate';
+import { getGeneratorInformation } from '../generate/generator-utils';
+import { VersionOptions } from './command-object';
+import { createNxReleaseConfig } from './config/config';
+import {
+  CATCH_ALL_RELEASE_GROUP,
+  ReleaseGroup,
+  createReleaseGroups,
+  handleCreateReleaseGroupsError,
+} from './config/create-release-groups';
+import { printDiff } from './utils/print-diff';
+
+export interface ReleaseVersionGeneratorSchema {
+  // The projects being versioned in the current execution
+  projects: ProjectGraphProjectNode[];
+  projectGraph: ProjectGraph;
+  specifier: string;
+  packageRoot?: string;
+  currentVersionResolver?: 'registry' | 'disk';
+  currentVersionResolverMetadata?: Record<string, unknown>;
+}
+
+export async function versionHandler(args: VersionOptions): Promise<void> {
+  const projectGraph = await createProjectGraphAsync({ exitOnError: true });
+  const nxJson = readNxJson();
+
+  const nxArgs = args as NxArgs;
+  if (nxArgs.verbose) {
+    process.env.NX_VERBOSE_LOGGING = 'true';
+  }
+
+  // Apply default configuration to any optional user configuration
+  const nxReleaseConfig = createNxReleaseConfig(nxJson.release);
+  const releaseGroupsData = await createReleaseGroups(
+    projectGraph,
+    nxReleaseConfig.groups
+  );
+  if (releaseGroupsData.error) {
+    return await handleCreateReleaseGroupsError(releaseGroupsData.error);
+  }
+
+  const tree = new FsTree(workspaceRoot, nxArgs.verbose);
+
+  let { releaseGroups } = releaseGroupsData;
+
+  /**
+   * User is filtering to a subset of projects. We need to make sure that what they have provided can be reconciled
+   * against their configuration in terms of release groups and the ungroupedProjectsHandling option.
+   */
+  if (args.projects?.length) {
+    const matchingProjectsForFilter = findMatchingProjects(
+      args.projects,
+      projectGraph.nodes
+    );
+
+    if (!matchingProjectsForFilter.length) {
+      output.error({
+        title: `Your --projects filter "${args.projects}" did not match any projects in the workspace`,
```

Comment:
I'm not convinced this should error. In the case of versioning within CI, it is possible that the output of `nx affected` is sent directly into this projects filter. In this case, it would make sense to gracefully exit with an info or warning log. 

Alternatively, we will need a way for CI to proactively determine whether or not the `nx release version` command should be run.

Reply:
This PR does not have versioning within CI as a design goal